prototypes/coin-selection/CoinSelection.hs

```diff
@@ -0,0 +1,278 @@
+--  * Maximum 'TxBody' size.
+--
+--      Usually in bytes, but this module does not prescribe a unit.
+--
+--  * Maximum 'TxOut' size.
+--
+--      For Cardano there are two maximums - one for
+--      serialized size of the 'TxOut' in bytes, and one for the maximum number of
+--      assets in a 'Value'.
+--
+--      Again, this module does not prescribe any particular unit.
+--
+--  * what is reward account balance?
+--
+-- __Optimization__
+--
+-- The coin selection function must minimize costs to the user. The following things should be part of the coin selection function:
+--
+--  * Fee minimization.
+--  * Use the smallest possible number of transactions to satisfy request.
+--  * Over time, in the long run, eventually, utxo distribution should evolve to match payment distribution (emergent utxo management).
+--
+-- __Algorithms__
+--
+--  * v'Sweep' - Select /all/ unspent coins, regardless of transaction outputs.
+--  * v'RandomImprove'
+--  * Random-Improve-MA - Multi-asset variant of Random-Improve.
+--
+
+-- | The transaction type is polymorphic and defined by the caller. Coin
+-- selection however needs to get certain information about the transaction
+-- under construction.
+--
+-- This class defines the types of transaction components and therefore
+-- represents transactions which can have the coin selection algorithms of this
+-- module applied to them.
+class TxBody tx where
+    -- | A transaction input - the source of funds for transaction.
+    type TxIn tx
+    -- | Payment destination for transaction.
+    type Address tx
+    -- | This type uniquely identifies a fungible asset.
+    type AssetId tx
+    -- | Quantity type of any given 'AssetId'.
+    --
+    -- The 'xyzzy' function requires this type to be a 'Monoid' under addition.
+    type Amount tx
+    -- | How fees are measured for this type of transaction. Its instance type
+    -- needn't necessarily be a monetary value.
+    type Cost tx
+
+    -- | Get the outputs of a 'TxBody'.
+    txOutputs :: tx -> [TxOut tx]
+
+    -- | Append an output to the 'TxBody'.
+    appendTxOutput :: TxOut tx -> tx -> tx
+
+    -- | Get the inputs of a 'TxBody'
+    txInputs :: tx -> [(TxIn tx, Value tx)]
+
+    -- | Add an input to the 'TxBody' input set.
+    addTxInput :: (TxIn tx, Value tx) -> tx -> tx
+
+-- | An abstract transaction output is defined as a payment destination and an
+-- amount.
+type TxOut tx = (Address tx, Value tx)
+
+{-
+-- | Possible values in a 'TxOut'.
+data TxOutValue tx
+    = MaxValue
+    -- | Take the maximum possible amount.
+    | Value (Value tx)
+    -- | Take a specific amount.
+    deriving Generic
+-}
+
+-- | A list of assets and their quantity.
+type Value tx = Map (AssetId tx) (Amount tx)
+
+-- | Unspent transaction outputs. These are the things which fund transactions.
+--
+-- Note that this type is a bit different from the usual map from TxIn -> TxOut.
+type UTxO tx = [(TxIn tx, Value tx)]
```

Comment:
We probably want to be able to handle the case of wallets with large `UTxO` sets, hence we probably need something with faster indexing, i.e. a `Map (TxIn tx) (Value tx)`.

Reply:
Yes, probably.
This type alias would be better named "AvailableInputs." There are no map lookups needed for coin selection.
Depending on the coin selection algorithm, `xyzzy` will either select randomly from these pairs, or sort them by value and take largest first.
So I renamed the element type, and changed the list to `Foldable`.